pkg/controller/volume/expand/cache/volume_resize_map.go

```diff
@@ -0,0 +1,285 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util/types"
+)
+
+// VolumeResizeMap defines an interface that serves as a cache for holding pending resizing requests
+type VolumeResizeMap interface {
+	AddPvcUpdate(newPvc *v1.PersistentVolumeClaim, oldPvc *v1.PersistentVolumeClaim, spec *volume.Spec)
+	GetPvcsWithResizeRequest() []*PvcWithResizeRequest
+	// Mark this volume as resize
+	MarkAsResized(*PvcWithResizeRequest) error
+	MarkForFileSystemResize(*PvcWithResizeRequest) error
+	MarkResizeFailed(*PvcWithResizeRequest, string) error
+}
+
+type volumeResizeMap struct {
+	// map of unique pvc name and resize requests that are pending or inflight
+	pvcrs map[types.UniquePvcName]*PvcWithResizeRequest
+	// kube client for making API calls
+	kubeClient clientset.Interface
+	// for guarding access to pvcrs map
+	sync.RWMutex
+}
+
+// PvcWithResizeRequest struct defines data structure that stores state needed for
+// performing file system resize
+type PvcWithResizeRequest struct {
```

Comment:
You are blurring functionality here. The purpose of the cache should be very crisp and clear: it stores in-memory state. It should not construct any requests, that should be the job of the reconciler, aka `sync_volume_resize.go`. In fact, the `volumeResizeMap struct` above should not even have a `kubeClient clientset.Interface` at all.

Reply:
I agree that `VolumeResizeMap` is more than just a cache of PVCs which needs to be resized, it also contains interface of functions which update the PVC after success/failure. 

There is certain simplicity to it, which makes testing easier. For example, if resize failed then the certain version of PVC must be added back to `VolumeResizeMap`. We can test that directly and it involves no goroutines and hence setting up the test scenario is simpler.  Every function call in `VolumeResizeMap` is direct and has no asynchronous behaviour. 

If we decouple updating PVC after resizing from `VolumeResizeMap` then - the reconciler will have to perform those updates. But it isn't as simple as that, because - sync_volume_resize calls volume operations through a goroutine (operation executor), it doesn't know about success or failure of particular operation, so in that case we will have to keep track of successfully resized PVCs separately.  So, we will have to introduce actual and desired state of world (or something along those lines).

But there is also inherent risk in keeping PVC objects in cache for too long (and in more than one place) - as jan pointed out, you may end up with stale version of PVC object and updating it may fail in etcd. 

The original version of code I wrote did use - actual & desired state of world but then I realized the design can be simplified if we just used one cache.